sae/blocks.go

```diff
@@ -104,74 +101,148 @@ func (vm *VM) buildBlock(
 		zap.Time("block_time", blockTime),
 	)
 
-	settleAt := unix(blockTime.Add(-params.Tau))
+	// The block time is not trusted, as it can be provided maliciously during
+	// [VM.VerifyBlock]. It is allowed for [hook.Points] to further constrain
+	// the allowed block times. However, every block MUST at least satisfy these
+	// basic sanity checks.
+	if blockUnix := blockTime.Unix(); blockUnix < saeparams.TauSeconds {
+		return nil, fmt.Errorf("%w: %d < %d", errBlockTimeUnderMinimum, blockUnix, saeparams.TauSeconds)
+	}
+	if parentTime := parent.Timestamp(); blockTime.Before(parentTime) {
+		return nil, fmt.Errorf("%w: %s < %s", errBlockTimeBeforeParent, blockTime, parentTime)
+	}
+	if maxBlockTime := vm.config.Now().Add(maxFutureBlockTime); blockTime.After(maxBlockTime) {
+		return nil, fmt.Errorf("%w: %s > %s", errBlockTimeAfterMaximum, blockTime, maxBlockTime)
+	}
+
+	settleAt := unix(blockTime.Add(-saeparams.Tau))
 	lastSettled, ok, err := blocks.LastToSettleAt(settleAt, parent)
 	if err != nil {
 		return nil, err
 	}
 	if !ok {
-		log.Warn("Execution lagging when determining last block to settle")
 		return nil, errExecutionLagging
 	}
 
-	// Although, at the time of writing this, the implementation of
-	// [blocks.LastToSettleAt] guarantees that the returned block has finished
-	// execution, this isn't a stated invariant and it's possible for it to be
-	// more aggressive in the future.
-	if !lastSettled.Executed() {
-		log.Warn(
-			"Execution lagging for settlement artefacts",
-			zap.Uint64("settling_height", lastSettled.Height()),
-		)
-		return nil, fmt.Errorf("%w: settling block %d: %T.WaitUntilExecuted(): %v", errExecutionLagging, lastSettled.Height(), lastSettled, err)
-	}
+	log = log.With(
+		zap.Uint64("last_settled_height", lastSettled.Height()),
+		zap.Stringer("last_settled_hash", lastSettled.Hash()),
+	)
 
-	sdb, err := state.New(lastSettled.PostExecutionStateRoot(), vm.exec.StateCache(), nil)
+	state, err := worstcase.NewState(vm.hooks, vm.exec.ChainConfig(), vm.exec.StateCache(), lastSettled)
 	if err != nil {
-		b := lastSettled
-		return nil, fmt.Errorf("state.New([post-execution state root %#x of block %d], ...): %v", b.PostExecutionStateRoot(), b.Height(), err)
-	}
-
-	// TODO(arr4n) worst-case tx validation will happen here. This is just a
-	// sketch of the process.
-	history := parent.WhenChildSettles(lastSettled)
-	var receipts types.Receipts
-	for _, b := range history {
-		receipts = append(receipts, b.Receipts()...)
+		log.Warn("Settled state not available")
```

Comment:
Is this in lieu of the `if !lastSettled.Executed()` check?

Reply:
The `.Executed()` check is performed internally (as that is where the execution artifacts are used).

I changed the log to refer to the worst-case state rather than the settled state, since that follows more from the actual code (and doesn't require knowledge of how worstcase is initialized)